### PR TITLE
Fix #875 Add Docker support when using VisualStudio 2017

### LINKED ISSUE #875: Add Docker support when using VisualStudio
Hello,

It would be great if we add docker support within Visual Studio 2017 using the best of the two worlds.

I can work on this task and submit PR.

This way we can build, debug and work with containers from inside VS2017. This will require only small changes and will make the solution more easy for new adopters.

Here is the list I am working on:

- [ ] Upgrade docker-compose files from 3.0 to 3.4
- [ ] Move Dockerfile image from 2.2-aspnetcore-runtime-alpine to 2.2-aspnetcore-runtime
- [ ] Modify .dockerignore (root level) to exclude .vs and .vscode recursively
- [ ] Modify .dockerignore to exclude non-sourcecode folders from build context (docs, templates, etc)
- [ ] Remove unnecessary sqlserver service in docker-compose-migrations.yml
- [ ] Add docker-compose project (.dcproj) to solution and configure to run migrations upon build
- [ ] Fix container published (host) port for all services in docker-compose-override.yaml to avoid dynamic ports to be added by Visual Studio 
- [ ] Add an exception to build-all.ps1 to specify fixed solution name when we have more than one sln/proj
- [ ] Add docker support for individual projects to be able to debug with right click -> debug -> new instance on containers
- [ ] Update documentation to new approach: split Microservices sample in different sections. (After PR #879  is reviewed)

Cheers

Bruno

### PR DESCRIPTION
**Pull request by brunobertechini**
Implemented required modifications to the Microservices Demo (Sample) to support Visual Studio. Optimized docker build context.

### CODE REVIEW

**Comment by brunobertechini** [comment]
@hikalkan please review the changes. In the meantime, I will try to check the documentation on abp.io for updates.

**Comment by hikalkan** [comment]
Thanks a lot @brunobertechini we will review it :)

**Comment by brunobertechini** [comment]
Thanks @hikalkan , Please let me know your thoughts and I can fix/enhance accordingly.

**Comment by brunobertechini** [comment]
@hikalkan , what is the reason behind not to add the actual references for Volo.Abp.* packages in the MSDemo solution ? I understand that the first time might be hard, and I also think that at the beginning you were adding too many projects...but it looks more stable now.

I am asking because I am studying the MS Demo and sometime I need to debug into Volo.Abp.* packages and I can't.

If you allow me, I can add all references in the solution. This will also remove the need to disable Nuget configuration to check for missing packages and manually restore inside VS2017 when opening solution.

**Comment by brunobertechini** [comment]
@atakanozceviz thanks for reviewing. Let me address your concerns one by one : 

> Why did you move images to 2.2-aspnetcore-runtime? Alpine images are much smaller, faster and more secure.

There is a bug on Visual Studio docker tools using alpine images. It relies on bash commands and other PID arguments not supported by alpine. This is meant to be fixed in VS2019. [See this Issue](https://developercommunity.visualstudio.com/content/problem/256756/new-container-tools-in-158-cant-debug-alpine-docke.html)

> Is there a specific reason to upgrade docker-compose files from 3.0 to 3.4? Why not 3.5 or 3.7? I think we should upgrade only when we really need it because everyone may not be able to access the latest version of Docker Engine and kompose don’t support anything 3.4 and above at the moment.

In order to support the target parameters in docker-compose, we need at least 3.4. Also this is the default version created by docker tools in visual studio. In ABP documentation, its stated that latest version of docker is required, so i do not think this is an issue.

Still, 3.4 is not the latest one.

image

> sqlserver service in docker-compose-migrations.yml is left intentionally to simplify restore database. 

The problem with this is that some docker engines (like the one used by VS and old versions of SWARM) may generate random port. Override should be used to specify changes (overrides). I do not see a problem of adding override all the times, since its required and should be part of standard compose run.

We can rever this if will cause problem, but this affects the behavior when you run all together. Maybe keep this and run migrations manually. 

With the current setup of this PR, VS2017 will always run the migrations automatically. As you can see in dcproj file.

> Host ports for sqlserver, mongodb, rabbitmq, redis, elasticsearch and logstash is not explicitly set to not conflict with the local development environment and the host ports for these services are not so important because we don't use them outside the docker network.

I think that most of developers will find it very helpfull to have fixed ports. We dont need to use the standard (same) ports of each service, but we can have them fixed. This way we can have consistent environments across machines.

> To enhance it is best to explicitly specifying port mappings as strings.

Good catch. This is certainly a required improvement. I will fix that.

Please let me know your thoughts

File: samples/MicroserviceDemo/docker-compose.override.yml [hunk 1]
```diff
@@ -60,7 +60,7 @@ services:
       - ReRoutes__2__DownstreamHostAndPorts__0__Port=80
       - GlobalConfiguration__BaseUrl=http://internal-gateway
     ports:
-      - "80"
+      - 51530:80
```
[the remark this answers is not in the document]
Comment (brunobertechini) [reply samples/MicroserviceDemo/docker-compose.override.yml]: Oh yes ! I missed the ports that were already there (originally in the source code). I will change them as well !

[the remark this answers is not in the document]
Comment (brunobertechini) [reply samples/MicroserviceDemo/docker-compose.override.yml]: @atakanozceviz done.

[the remark this answers is not in the document]
Comment (atakanozceviz) [reply samples/MicroserviceDemo/docker-compose.override.yml]: @brunobertechini Thank you!

**Comment by atakanozceviz** [comment]
Hi @brunobertechini , I am trying to run the project using docker-compose in Visual Studio, unfortunately I'm getting `System.Net.Http.HttpRequestException` exception from BackendAdminApp.Host and PublicWebSite.Host. Do you get the same error? 
It is working fine when I run the project from command line using docker-compose.

**Pull request closed without merging**